internal/state/resolver/resolver.go

```diff
@@ -142,13 +149,53 @@ func endpointReady(endpoint discoveryV1.Endpoint) bool {
 	return ready != nil && *ready
 }
 
-func targetPortExists(ports []discoveryV1.EndpointPort, targetPort int32) bool {
-	for _, port := range ports {
-		// A nil port indicates that all ports apply to this EndpointSlice.
-		if port.Port == nil || *port.Port == targetPort {
-			return true
+func filterEndpointSliceList(
+	endpointSliceList discoveryV1.EndpointSliceList,
+	port v1.ServicePort,
+) []discoveryV1.EndpointSlice {
+	filtered := make([]discoveryV1.EndpointSlice, 0, len(endpointSliceList.Items))
+
+	for _, endpointSlice := range endpointSliceList.Items {
+		if !ignoreEndpointSlice(endpointSlice, port) {
+			filtered = append(filtered, endpointSlice)
+		}
+	}
+
+	return filtered
+}
+
+// findPort locates the endpoint port for the given EndpointPort slice and ServicePort.
+// If the targetPort of the ServicePort is a string, we find the endpoint port by name. If the targetPort of the
+// ServicePort is an int, we find the endpoint port by value. If the endpoint port is nil, then according to the spec,
+// all ports should be considered a match. In this case, we return the ServicePort port value for string targetPorts,
+// or the targetPort integer value. If no match is found, return 0.
+func findPort(ports []discoveryV1.EndpointPort, svcPort v1.ServicePort) int32 {
```

Comment:
I caught the following bug:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: coffee
spec:
  replicas: 1
  selector:
    matchLabels:
      app: coffee
  template:
    metadata:
      labels:
        app: coffee
    spec:
      containers:
      - name: coffee
        image: nginxdemos/nginx-hello:plain-text
        ports:
        - containerPort: 8080
          name: myport
---
apiVersion: v1
kind: Service
metadata:
  name: coffee
spec:
  ports:
  - port: 80
    targetPort: myport
    protocol: TCP
    name: http
  selector:
    app: coffee
```    

EndpointSlice ports:
```yaml
ports:
- name: http
  port: 8080
  protocol: TCP
```

logs:
```
{"level":"info","ts":1664299471.3182275,"logger":"changeProcessor","msg":"Got warning while building graph","kind":"HTTPRoute","namespace":"default","name":"coffee","warning":"cannot resolve backend ref for rule 0: no valid endpoints found for Service default/coffee and port {http TCP <nil> 80 {1 0 myport} 0}"}
```

expected - no error.

I also wonder if it make sense to print the name of the fields for the port, so that it would be easier to parse it for humans in the logs?

Reply:
🤦‍♀️ I should've tested with a service port name that was different from the target port name. Thanks for catching. 

So the k8s EndpointSlice controller is already doing the mapping of target port to container port. I just need to match the service port name to the endpoint port name.

However, the service port name is optional if there's only one service port defined. 

I need to do more testing.

Reply:
Confirmed that the Kubernetes EndpointSlice controller is handling all of the logic of mapping target port to container port (see this [function](https://github.com/kubernetes/kubernetes/blob/5835544ca568b757a8ecae5c153f317e5736700e/pkg/controller/endpointslice/utils.go#L85)).

I updated `findPort` to return the EndpointPort port that matches the ServicePort name. In the case where the ServicePort is unnamed, we can still use this logic because the corresponding EndpointPort port will also be unnamed (unnamed is defined as a name with an empty string value).